python/wtf-python/default-mutable-dict.yaml

```diff
@@ -0,0 +1,184 @@
+                $D = $D.copy()
+                ...
+                $D.update(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = copy.deepcopy($D)
+                ...
+                $D.update(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = copy.copy($D)
+                ...
+                $D.update(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = dict.copy($D)
+                ...
+                $D.update(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = {... for ... in ...}
+                ...
+                $D.update(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = $D or {}
+                ...
+                $D.update(...)
+          - pattern-not-inside: |
+              def $A(...):
+                ...
+                def $F(..., $D={}, ...):
+                  ...
+                  $D.update(...)
+        - patterns:
+          - pattern: |
+              def $F(..., $D={}, ...):
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = {}
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = dict(...)
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = $D.copy()
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = copy.deepcopy($D)
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = copy.copy($D)
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = dict.copy($D)
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = {... for ... in ...}
+                ...
+                $D.setdefault(...)
+          - pattern-not: |
+              def $F(..., $D={}, ...):
+                ...
+                $D = $D or {}
+                ...
+                $D.setdefault(...)
+          - pattern-not-inside: |
+              def $A(...):
+                ...
+                def $F(..., $D={}, ...):
+                  ...
+                  $D.setdefault(...)
+    message: "Function $F mutates default dict $D"
```

Comment:
I'm not sure if people will immediately see _why_ this is bad. I might suggest adding:

> Default function arguments in Python are shared between function calls. Mutating a default function argument changes the result for all function consumers. This can cause unexpected results, or lead to security vulnerabilities whereby one function consumer can either view or modify the data of another function consumer.

Reply:
Even this is not enough IMO. We need to link to a blog post with examples. Do we want to wordsmith error messages here or once a check has been brought into the golden set?

Reply:
Typically what I'll do is get the check message close to good enough, because we might forget to edit the check messages before they get shipped 🤷‍♂️ . No need to have any back-and-forths at this stage though

Reply:
Made the suggested change.